File: executor/src/main/java/com/codeloom/executor/service/SubmissionProcessingService.java
package com.codeloom.executor.service;

import static com.codeloom.executor.engine.CodeExecutionConstants.*;

import com.codeloom.common.*;
import com.codeloom.common.event.*;
import com.codeloom.common.language.LanguageSpec;
import com.codeloom.executor.engine.*;
import com.codeloom.executor.repository.TestCaseRepository;
import java.util.*;
import org.slf4j.*;
import org.springframework.stereotype.Service;

@Service
public class SubmissionProcessingService {
  private static final Logger log = LoggerFactory.getLogger(SubmissionProcessingService.class);
  private final TestCaseRepository tests;
  private final DockerJudgeEngine judge;
  private final EventService events;

  public SubmissionProcessingService(TestCaseRepository t, DockerJudgeEngine j, EventService e) {
    tests = t;
    judge = j;
    events = e;
  }

  public void process(SubmissionEvent e) {
    var cases = tests.findByProblemId(e.problemId());
    var c =
        new SubmissionContext(
            e.submissionId(),
            e.userId(),
            e.problemId(),
            e.code(),
            LanguageSpec.fromLanguage(e.language()),
            e.executionTimeLimitMs(),
            e.memoryUsageLimitBytes());
    if (cases.isEmpty()) {
      changeSubmissionStatus(c, SubmissionStatus.ACCEPTED);
      return;
    }
    List<TestCaseResult> results = new ArrayList<>();
    try {
      changeSubmissionStatus(c, SubmissionStatus.COMPILING);
      if (!judge.compile(c).isSuccessful()) {
        changeSubmissionStatus(c, SubmissionStatus.COMPILE_ERROR);
        return;
      }
      changeSubmissionStatus(c, SubmissionStatus.RUNNING);
      for (var t : cases) {
        RunResult r = judge.runTestCase(c, t);
        if (t.isPublic())
          results.add(
              new TestCaseResult(
                  t.getId(),
                  t.getProblemId(),
                  t.getInput(),
                  t.getExpectedOutput(),
                  r.stdout(),
                  r.stderr(),
                  r.executionTimeMs(),
                  r.memoryUsageBytes()));
        if (r.exitCode() != 0) {
          SubmissionStatus s =
              r.exitCode() == TIMEOUT_EXIT_CODE
                  ? SubmissionStatus.TIME_LIMIT_EXCEEDED
                  : r.exitCode() == MEMORY_LIMIT_EXCEEDED_EXIT_CODE
                      ? SubmissionStatus.MEMORY_LIMIT_EXCEEDED
                      : SubmissionStatus.RUNTIME_ERROR;
          changeSubmissionStatus(c, s, new SubmissionStatusPayload(null, results));
          return;
        }
        if (!r.stdout().trim().equals(t.getExpectedOutput().trim())) {
          changeSubmissionStatus(
              c, SubmissionStatus.WRONG_ANSWER, new SubmissionStatusPayload(null, results));
          return;
        }
      }
      changeSubmissionStatus(
          c, SubmissionStatus.ACCEPTED, new SubmissionStatusPayload(null, results));
    } catch (Exception x) {
      log.error("Error while processing submission={}", c, x);
      changeSubmissionStatus(
          c, SubmissionStatus.SYSTEM_ERROR, new SubmissionStatusPayload(null, results));
    } finally {
      judge.cleanup(c.submissionId());
    }
  }

  public void changeSubmissionStatus(SubmissionContext c, SubmissionStatus s) {
    changeSubmissionStatus(c, s, null);
  }

  public void changeSubmissionStatus(
      SubmissionContext c, SubmissionStatus s, SubmissionStatusPayload p) {
    log.info("Submission(id={}) status changed to {}", c.submissionId(), s);
    events.submissionStatusChanged(c, s, p);
  }
}

File: executor/src/main/kotlin/com/codeloom/executor/service/SubmissionProcessingService.kt
package com.codeloom.executor.service

import com.codeloom.common.SubmissionEvent
import com.codeloom.common.SubmissionStatus
import com.codeloom.common.event.SubmissionStatusPayload
import com.codeloom.common.event.TestCaseResult
import com.codeloom.executor.engine.DockerJudgeEngine
import com.codeloom.executor.engine.MEMORY_LIMIT_EXCEEDED_EXIT_CODE
import com.codeloom.executor.engine.SubmissionContext
import com.codeloom.executor.engine.TIMEOUT_EXIT_CODE
import com.codeloom.common.language.LanguageSpec
import com.codeloom.executor.repository.TestCaseRepository
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service

@Service
class SubmissionProcessingService(
    private val testCaseRepository: TestCaseRepository,
    private val dockerJudgeEngine: DockerJudgeEngine,
    private val eventService: EventService,
) {
    private val logger = LoggerFactory.getLogger(SubmissionProcessingService::class.java)

    fun process(event: SubmissionEvent) {
        val testCases = testCaseRepository.findByProblemId(event.problemId)
        val context =
            SubmissionContext(
                submissionId = event.submissionId,
                userId = event.userId,
                problemId = event.problemId,
                code = event.code,
                language = LanguageSpec.fromLanguage(event.language),
                executionTimeLimitMs = event.executionTimeLimitMs,
                memoryUsageLimitBytes = event.memoryUsageLimitBytes,
            )

        if (testCases.isEmpty()) {
            logger.warn("Problem doesn't have any test cases: problemId={}", event.problemId)
            changeSubmissionStatus(context, SubmissionStatus.ACCEPTED)
            return
        }

        val testCaseResults = mutableListOf<TestCaseResult>()
        try {
            changeSubmissionStatus(context, SubmissionStatus.COMPILING)
            val compilationResult = dockerJudgeEngine.compile(context)
            if (!compilationResult.isSuccessful) {
                changeSubmissionStatus(context, SubmissionStatus.COMPILE_ERROR)
                return
            }

            changeSubmissionStatus(context, SubmissionStatus.RUNNING)
            for (testCase in testCases) {
                val runResult = dockerJudgeEngine.runTestCase(context, testCase)

                if (testCase.isPublic) {
                    testCaseResults +=
                        TestCaseResult(
                            id = testCase.id!!,
                            problemId = testCase.problemId,
                            input = testCase.input,
                            expectedOutput = testCase.expectedOutput,
                            stdout = runResult.stdout,
                            stderr = runResult.stderr,
                            executionTimeMs = runResult.executionTimeMs,
                            memoryUsageBytes = runResult.memoryUsageBytes,
                        )
                }

                if (runResult.exitCode != 0L) {
                    val newStatus =
                        when (runResult.exitCode) {
                            TIMEOUT_EXIT_CODE -> SubmissionStatus.TIME_LIMIT_EXCEEDED
                            MEMORY_LIMIT_EXCEEDED_EXIT_CODE -> SubmissionStatus.MEMORY_LIMIT_EXCEEDED
                            else -> SubmissionStatus.RUNTIME_ERROR
                        }
                    changeSubmissionStatus(
                        context,
                        newStatus,
                        SubmissionStatusPayload(testCaseResults = testCaseResults),
                    )
                    return
                }

                if (runResult.stdout != testCase.expectedOutput) {
                    changeSubmissionStatus(
                        context,
                        SubmissionStatus.WRONG_ANSWER,
                        SubmissionStatusPayload(testCaseResults = testCaseResults),
                    )
                    return
                }
            }
            changeSubmissionStatus(
                context,
                SubmissionStatus.ACCEPTED,
                SubmissionStatusPayload(testCaseResults = testCaseResults),
            )
        } catch (e: Exception) {
            logger.error("Error while processing submission={}", context, e)
            changeSubmissionStatus(
                context,
                SubmissionStatus.SYSTEM_ERROR,
                SubmissionStatusPayload(testCaseResults = testCaseResults),
            )
        } finally {
            dockerJudgeEngine.cleanup(context.submissionId)
        }
    }

    fun changeSubmissionStatus(
        context: SubmissionContext,
        status: SubmissionStatus,
        payload: SubmissionStatusPayload? = null,
    ) {
        logger.info("Submission(id={}) status changed to {}", context.submissionId, status)
        eventService.submissionStatusChanged(context, status, payload)
    }
}
